src/auth/auth.controller.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  Controller,
+  Get,
+  Query,
+  Req,
+  Res,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { validateSync } from 'class-validator';
+import { plainToInstance } from 'class-transformer';
+import { Request, Response } from 'express';
+import { AuthService } from './auth.service';
+import { SgidCallbackCookieDto } from './auth.dto';
+import { SgidAuthStatus } from './auth.constants';
+
+const SGID_PO_COOKIE_NAME = 'SGID_PO_COOKIE_NAME';
+
+@Controller('auth')
+export class AuthController {
+  constructor(private readonly authService: AuthService) {}
+  @Get('sgid/auth-url')
+  generateSgIdAuthUrl(
+    @Query('chatId') chatId: string,
+    @Res({ passthrough: true }) res: Response,
+  ): string {
+    const { codeVerifier, codeChallenge, nonce } =
+      this.authService.generateSgidLoginParams();
+    const url = this.authService.createSgidAuthUrl({
+      chatId,
+      codeChallenge,
+      nonce,
+    });
+    const cookie = JSON.stringify({
+      codeVerifier,
+      nonce,
+    });
+
+    res.cookie(SGID_PO_COOKIE_NAME, cookie, { httpOnly: true });
+    res.redirect(url);
+    return url;
+  }
+
+  @Get('sgid/callback')
+  async sgidCallback(
+    @Req() req: Request,
+    @Query('code') code: string,
+    @Res({ passthrough: true }) res: Response,
+  ) {
+    // One time validity
+    res.clearCookie(SGID_PO_COOKIE_NAME);
+
+    const rawCookie = req.cookies[SGID_PO_COOKIE_NAME];
+    if (typeof rawCookie !== 'string') {
+      throw new UnauthorizedException(
+        "You're not authorised to access this link.",
+      );
+    }
+
+    // Extract & Validate cookie
+    const cookie = JSON.parse(rawCookie);
+    const cookieInstance = plainToInstance(SgidCallbackCookieDto, cookie);
+    const errors = validateSync(cookieInstance);
+    if (errors.length > 0) {
+      throw new UnauthorizedException(
+        "You're not authorised to access this link.",
+      );
+    }
+
+    const authStatus = await this.authService.verifyUserFromAuthCode({
+      code,
+      nonce: cookieInstance.nonce,
+      codeVerifier: cookieInstance.codeVerifier,
+    });
+
+    // TODO: Insert custom flow depending on the authentication status.
+    return SgidAuthStatus[authStatus];
```

Comment:
Hmm. Would be nice if we could 302 redirect it back to the bot at `https://t.me/<botname>` (like how the CPF app would open up Singpass login in an in-app browser view, then prompt you to go back to the app). The bot could then send a message depending on the auth outcome, perhaps invoked from this controller.

If you want a demonstration of how this works in the CPF app come find me later.

Also, why are you using an enum to index into the same enum?

Reply:
Yeap, I'll give it a go!

sorry I was thinking to reflect the Enum text on FE during development 😅